frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+}
+
+function createNotificationData (
+  notificationType: string,
+  moreParams: Object = {}
+): Object {
+  return {
+    type: MESSAGE_TYPES.NOTIFICATION,
+    notification: {
+      type: notificationType,
+      ...moreParams
+    }
+  }
+}
+
+function emitMessageEvents ({ type, contractID, hash, state }: {
+  type: string, contractID: string, hash: string, state: Object
+}): void {
+  for (let i = state.messages.length - 1; i >= 0; i--) {
+    if (state.messages[i].id === hash) {
+      sbp('okTurtles.events/emit', `${CHATROOM_MESSAGE_ACTION}-${contractID}`, {
+        type,
+        data: { message: state.messages[i] }
+      })
+      break
+    }
+  }
+}
+
+sbp('chelonia/defineContract', {
+  name: 'gi.contracts/chatroom',
+  metadata: {
+    validate: objectOf({
+      createdDate: string, // action created date
+      username: string, // action creator
+      identityContractID: string // action creator identityContractID
+    }),
+    create () {
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: chatRoomType,
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
+          settings: {
+            messagesPerPage: CHATROOM_MESSAGES_PER_PAGE,
+            maxNameLetters: CHATROOM_NAME_LIMITS_IN_CHARS,
+            maxDescriptionLetters: CHATROOM_DESCRIPTION_LIMITS_IN_CHARS
+          },
+          users: {},
+          messages: []
+        }, {
+          attributes: {
+            ...data,
+            creator: meta.username,
+            deletedDate: null,
+            archievedDate: null
+          }
+        })
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
+        }
+      },
+      sideEffect ({ contractID, meta }) {
+        const rootState = sbp('state/vuex/state')
+        if (meta.username === rootState.loggedIn.username) {
+          sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
+        }
+      }
```

Comment:
I'm not entirely sure about this...

What if we've just joined a group with multiple chat contracts to sync? Is this going to then change the `currentChatRoomId` multiple times, and then set it to something random (whichever chatroom we sync last)?

Reply:
When we join a group, we only join the `General` chatroom by default.
Group chatrooms are listed in the `chatRooms` of the group contract, so even though a user has not joined(synced) all the channels, he can see the channels list.
This redirection is needed when a user creates a new chatroom making him redirect to the newly-created chatroom automatically.

Reply:
Ah, true.

A different question then: why it it wrapped inside of `if (meta.username === rootState.loggedIn.username) {`?

Shouldn't we join the general chatroom even if we weren't the user who created it? (since `meta.username` = the user who created the contract...)

Reply:
Yeap. We need to join `General` chatroom automatically if we join any group.
The following shows how we automatically join the `General` chatroom when we join a group.
```
'gi.actions/group/join': async function (params: $Exact<GIActionParams>) {
    try {
      // post acceptance event to the group contract
      const message = await sbp('chelonia/out/actionEncrypted', {
        action: 'gi.contracts/group/inviteAccept', ...params
      })
      // sync the group's contract state
      await sbp('state/enqueueContractSync', params.contractID)

      // join the 'General' chatroom by default
      const rootState = sbp('state/vuex/state')
      const generalChatRoomId = rootState[params.contractID].generalChatRoomId
      if (generalChatRoomId) {
        await sbp('gi.actions/group/joinChatRoom', {
          contractID: params.contractID,
          data: { chatRoomID: generalChatRoomId }
        })
      }

      return message
    } catch (e) {
      console.error('gi.actions/group/join failed!', e)
      throw new GIErrorUIRuntimeError(L('Failed to join the group: {codeError}', { codeError: e.message }))
    }
  }
```
The condition `(meta.username === rootState.loggedIn.username)` is to redirect channel creators to the newly-created channel only when they create channels.

Reply:
Ok, can you put what you just wrote into a comment above that code?

> redirect channel creators to the newly-created channel only when they create channels.

Reply:
Okay.